src/operations/Operation.cpp

```diff
@@ -121,6 +121,58 @@ bool Operation::equals(const Operation& op, const Permutation& perm1,
     return false;
   }
 
+  if (isDiagonalGate()) {
+    // check pos. controls and targets together
+    std::set<Qubit> usedQubits1{};
+    std::set<Qubit> negControls1{};
+    if (perm1.empty()) {
+      usedQubits1 = {targets.cbegin(), targets.cend()};
+      for (const auto& control : controls) {
+        if (control.type == Control::Type::Pos) {
+          usedQubits1.emplace(control.qubit);
+        } else {
+          negControls1.emplace(control.qubit);
+        }
+      }
+    } else {
+      for (const auto& control : controls) {
+        if (control.type == Control::Type::Pos) {
+          usedQubits1.emplace(perm1.at(control.qubit));
+        } else {
+          negControls1.emplace(perm1.at(control.qubit));
+        }
+      }
+      for (const auto& qubit : targets) {
+        usedQubits1.emplace(perm1.at(qubit));
+      }
+    }
+
+    std::set<Qubit> usedQubits2{};
+    std::set<Qubit> negControls2{};
+    if (perm2.empty()) {
+      usedQubits2 = {op.getTargets().cbegin(), op.getTargets().cend()};
+      for (const auto& control : op.getControls()) {
+        if (control.type == Control::Type::Pos) {
+          usedQubits2.emplace(control.qubit);
+        } else {
+          negControls2.emplace(control.qubit);
+        }
+      }
+    } else {
+      for (const auto& control : op.getControls()) {
+        if (control.type == Control::Type::Pos) {
+          usedQubits2.emplace(perm2.at(control.qubit));
+        } else {
+          negControls2.emplace(perm2.at(control.qubit));
+        }
+      }
+      for (const auto& qubit : op.getTargets()) {
+        usedQubits2.emplace(perm2.at(qubit));
+      }
+    }
+
+    return usedQubits1 == usedQubits2 && negControls1 == negControls2;
```

Comment:
Just an idea at the top of my head:
Quite some code in this function is simply duplicated more-or-less four times. two variants for the two operations and two variants for permutation or no permutation.
Guess it would be possible to simplify this a little bit by introducing further protected member functions in the `Operation` class that cover parts of the code here.
Probably too much to ask for this PR, so feel free to defer to a later point in time. Just wanted to bring it up.

I was also wondering: there is already a `getUsedQubits` method. Could it be worth reusing that here? Maybe in addition to a new `getCointrolType() -> std::vector<qc::Control::Type>` or `getControlPolarity() -> std::vector<bool>` member function.

Reply:
That does look better already! 👍🏻
Could someone use a lambda function that is either the identity or applies the permutation and use such a function to construct the respective sets?
One could, for example, augment the `getUsedQubits` function with an optional std::function argument that defaults to the identity.

I believe that could help improve the clarity throughout the whole function here even more.

Reply:
Realized that, I think now it is easier when you look a the "files changed" tab.